test/lit/passes/gto-mutability.wast

```diff
@@ -688,3 +688,327 @@
+;; fields.
+(module
+  (rec
+    ;; CHECK:      (rec
+    ;; CHECK-NEXT:  (type $super (sub (struct (field i32))))
+    (type $super (sub (struct (field (mut i32)))))
+    ;; CHECK:       (type $mid (sub $super (struct (field i32) (field (ref string)))))
+    (type $mid (sub $super (struct (field (mut i32)) (field (mut (ref string))))))
+    ;; CHECK:       (type $sub (sub $mid (struct (field i32) (field (ref string)))))
+    (type $sub (sub $mid (struct (field (mut i32)) (field (mut (ref string))))))
+  )
+
+  ;; CHECK:       (type $3 (func))
+
+  ;; CHECK:      (func $test (type $3)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.get $sub 1
+  ;; CHECK-NEXT:    (struct.new $sub
+  ;; CHECK-NEXT:     (i32.const 42)
+  ;; CHECK-NEXT:     (string.const "foo")
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.get $mid 1
+  ;; CHECK-NEXT:    (struct.new $mid
+  ;; CHECK-NEXT:     (i32.const 1337)
+  ;; CHECK-NEXT:     (string.const "bar")
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.get $super 0
+  ;; CHECK-NEXT:    (struct.new $super
+  ;; CHECK-NEXT:     (i32.const 999999)
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $test
+    (drop
+      (struct.get $sub 1
+        (struct.new $sub
+          (i32.const 42)
+          (string.const "foo")
+        )
+      )
+    )
+    (drop
+      (struct.get $mid 1
+        (struct.new $mid
+          (i32.const 1337)
+          (string.const "bar")
+        )
+      )
+    )
+    ;; Only a get of the first field.
+    (drop
+      (struct.get $super 0
+        (struct.new $super
+          (i32.const 999999)
+        )
+      )
+    )
+  )
+)
+
+;; The super is public, but we can still optimize the field in the sub.
+(module
+  ;; CHECK:      (type $super (sub (struct)))
+  (type $super (sub (struct)))
+
+  ;; CHECK:      (rec
+  ;; CHECK-NEXT:  (type $sub (sub $super (struct (field stringref))))
+  (type $sub (sub $super (struct (field (mut stringref)))))
+
+  ;; CHECK:       (type $2 (func))
+
+  ;; CHECK:      (global $global (ref $super) (struct.new_default $super))
+  (global $global (ref $super) (struct.new_default $super))
+  ;; CHECK:      (export "global" (global $global))
+  (export "global" (global $global))
+
+  ;; CHECK:      (func $test (type $2)
+  ;; CHECK-NEXT:  (drop
+  ;; CHECK-NEXT:   (struct.get $sub 0
+  ;; CHECK-NEXT:    (struct.new $sub
+  ;; CHECK-NEXT:     (string.const "foo")
+  ;; CHECK-NEXT:    )
+  ;; CHECK-NEXT:   )
+  ;; CHECK-NEXT:  )
+  ;; CHECK-NEXT: )
+  (func $test
+    ;; Write and read the field.
+    (drop
+      (struct.get $sub 0
+        (struct.new $sub
+          (string.const "foo")
+        )
+      )
+    )
+  )
+)
+
+;; As above, and now the super has the field as well, preventing optimization.
```

Comment:
Do we still allow refinement of fields in private subtypes whose supertypes are public? That should be fine. I guess this code path only handles mutability, though.

Reply:
Yes (when they are immutable), in TypeRefining, see the last two tests here:

https://github.com/WebAssembly/binaryen/blob/main/test/lit/passes/type-refining.wast#L1387-L1522

Reply:
Aha, it's a separate pass, thanks. Is there a specific reason these optimizations are separate? It seems like they should have a lot in common.

Reply:
They do share much logic through StructUtils, TypeUpdater, etc. The mutability part could have perhaps been moved over to TypeRefining (in contrast to field reordering and removal), that might be for historical reasons.